ndt7/receiver/receiver.go

```diff
@@ -19,28 +20,44 @@ type receiverKind int
 const (
 	downloadReceiver = receiverKind(iota)
 	uploadReceiver
+	pingReceiver
+)
+
+const (
+	MaxDuration = math.MaxInt64 * time.Nanosecond
 )
 
 func loop(
 	ctx context.Context, conn *websocket.Conn, kind receiverKind,
-	dst chan<- model.Measurement,
+	dst chan<- model.Measurement, start time.Time, pongch chan<- model.WSInfo,
 ) {
 	logging.Logger.Debug("receiver: start")
 	defer logging.Logger.Debug("receiver: stop")
 	defer close(dst)
+	defer close(pongch)
 	conn.SetReadLimit(spec.MaxMessageSize)
 	receiverctx, cancel := context.WithTimeout(ctx, spec.MaxRuntime)
 	defer cancel()
-	err := conn.SetReadDeadline(time.Now().Add(spec.MaxRuntime)) // Liveness!
+	err := conn.SetReadDeadline(start.Add(spec.MaxRuntime)) // Liveness!
 	if err != nil {
 		logging.Logger.WithError(err).Warn("receiver: conn.SetReadDeadline failed")
 		return
 	}
+	minRTT := MaxDuration
 	conn.SetPongHandler(func(s string) error {
-		rtt, err := ping.ParseTicks(s)
+		elapsed, rtt, err := ping.ParseTicks(s, start)
 		if err == nil {
-			rtt /= int64(time.Millisecond)
-			logging.Logger.Debugf("receiver: ApplicationLevel RTT: %d ms", rtt)
+			logging.Logger.Debugf("receiver: ApplicationLevel RTT: %d ms", int64(rtt / time.Millisecond))
+			if rtt < minRTT {
+				minRTT = rtt
+			}
+
+			wsinfo := model.WSInfo{
+				ElapsedTime: int64(elapsed / time.Microsecond),
+				LastRTT: int64(rtt / time.Microsecond),
+				MinRTT: int64(minRTT / time.Microsecond),
+			}
+			pongch <- wsinfo // Liveness: buffered (sender)
```

Comment:
So, here I believe it should be possible to use `dst` to emit a measurement containing PingInfo, right?

Reply:
Yes, it may be written to the resulting json file here, it will be put under `ClientMeasurements` in this case. But `pongch` is still useful to send the message back to ndt7 client for the liveness reasons mentioned above.